rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -1215,21 +1215,96 @@ RMW_Connext_Subscriber::create(
     });
 
   DDS_TopicDescription * sub_topic = DDS_Topic_as_topicdescription(topic);
-
-  if (nullptr != cft_name) {
-    rmw_ret_t cft_rc =
-      rmw_connextdds_create_contentfilteredtopic(
-      ctx, dp, topic, cft_name, cft_filter, &cft_topic);
+  std::string cft_topic_name;
+  if (nullptr != cft_name || nullptr != subscriber_options->filter_expression) {
+    rmw_ret_t cft_rc = RMW_RET_OK;
+
+    if (nullptr != cft_name) {
+      cft_rc =
+        rmw_connextdds_create_contentfilteredtopic(
+        ctx, dp, topic, cft_name, cft_filter, NULL, &cft_topic);
+    } else if (nullptr != subscriber_options->filter_expression) {
+      cft_topic_name =
+        fqtopic_name + "_ContentFilterTopic" + RMW_Connext_Subscriber::get_atomic_id();
+      cft_rc =
+        rmw_connextdds_create_contentfilteredtopic(
+        ctx, dp, topic, cft_topic_name.c_str(),
+        subscriber_options->filter_expression,
+        subscriber_options->expression_parameters, &cft_topic);
+    }
 
     if (RMW_RET_OK != cft_rc) {
       if (RMW_RET_UNSUPPORTED != cft_rc) {
         return nullptr;
       }
     } else {
-      sub_topic = cft_topic;
+      if (nullptr != cft_topic) {
+        sub_topic = cft_topic;
+      }
     }
   }
 
+  DDS_DataReader * dds_reader = initialize_datareader(
+    ctx,
+    dp,
+    sub,
+    fqtopic_name,
+    qos_policies,
+    type_support,
+    subscriber_options,
+    internal,
+    sub_topic);
+
+  auto scope_exit_dds_reader_delete =
+    rcpputils::make_scope_exit(
+    [sub, dds_reader]()
+    {
+      if (DDS_RETCODE_OK !=
+      DDS_Subscriber_delete_datareader(sub, dds_reader))
+      {
+        RMW_CONNEXT_LOG_ERROR_SET(
+          "failed to delete DDS DataWriter")
+      }
+    });
+
+  RMW_Connext_Subscriber * rmw_sub_impl =
+    new (std::nothrow) RMW_Connext_Subscriber(
+    ctx,
+    dds_reader,
+    topic,
+    type_support,
+    subscriber_options->ignore_local_publications,
+    topic_created,
+    cft_topic,
+    internal);
+
+  if (nullptr == rmw_sub_impl) {
+    RMW_CONNEXT_LOG_ERROR_SET("failed to allocate RMW subscriber")
+    return nullptr;
+  }
+  rmw_sub_impl->fqtopic_name = fqtopic_name;
+  rmw_sub_impl->qos_policies = *qos_policies;
+  rmw_sub_impl->subscriber_options = *subscriber_options;
```

Comment:
I would prefer if these fields were passed directly to the constructor, and cached inside, for considerations similar to the one I mentioned about the `node` fields (making it clear that these are values determined at creation time and which will remain unchanged for the duration of the object's lifetime).

It will also help us keep track of them if they ever develop from "simple data" which doesn't require explicit finalization, to some more complex structure with their own `finalize()` function to call.

Reply:
Updated